src/main/java/com/openlattice/datastore/data/controllers/DataController.java

```diff
@@ -981,4 +991,41 @@ private static OffsetDateTime getDateTimeFromLong( long epochTime ) {
         return OffsetDateTime.ofInstant( Instant.ofEpochMilli( epochTime ), ZoneId.systemDefault() );
     }
 
+    private Stream<UUID> streamEntitySetIds( DataAssociation association ) {
+        return Stream.of( association.getSrcEntitySetId(), association.getDstEntitySetId() );
+    }
+
+    private Stream<UUID> streamEntitySetIds( DataEdge dataEdge ) {
+        return Stream.of( dataEdge.getSrc().getEntitySetId(), dataEdge.getDst().getEntitySetId() );
+    }
+
+    private Stream<UUID> streamEntitySetIds( DataEdgeKey dataEdgeKey ) {
+        return Stream.of( dataEdgeKey.getEdge().getEntitySetId(),
+                dataEdgeKey.getSrc().getEntitySetId(),
+                dataEdgeKey.getDst().getEntitySetId() );
+    }
+
+    private <T> Set<UUID> getEntitySetIdsFromCollection(
+            Collection<T> items,
+            Function<T, Stream<UUID>> transformation ) {
+        return items.stream().flatMap( transformation ).collect( Collectors.toSet() );
+    }
+
+    private void checkPermissionsOnEntitySetIds( Set<UUID> entitySetIds, EnumSet<Permission> permissions ) {
+        //Ensure that we have write access to entity sets.
+        ensureEntitySetsCanBeWritten( entitySetIds );
+        accessCheck( entitySetIds.stream().collect( Collectors.toMap( AclKey::new, id -> permissions ) ) );
+    }
+
+    private void ensureEntitySetCanBeWritten( UUID entitySetId ) {
+        ensureEntitySetsCanBeWritten( ImmutableSet.of( entitySetId ) );
+    }
+
+    private void ensureEntitySetsCanBeWritten( Set<UUID> entitySetIds ) {
+        if ( entitySetService.entitySetsContainFlag( entitySetIds, EntitySetFlag.AUDIT ) ) {
+            throw new ForbiddenException( "You cannot modify data of entity sets " + entitySetIds.toString()
```

Comment:
I would think we'd want to differentiate entity sets that do have the flag vs those that don't but maybe that's TMI

Reply:
+1